src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java

```diff
@@ -0,0 +1,171 @@
+import org.prebid.server.bidder.model.BidderBid;
+import org.prebid.server.bidder.model.BidderError;
+import org.prebid.server.bidder.model.HttpCall;
+import org.prebid.server.bidder.model.HttpRequest;
+import org.prebid.server.bidder.model.Result;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.proto.openrtb.ext.request.applogy.ExtImpApplogy;
+import org.prebid.server.proto.openrtb.ext.response.BidType;
+import org.prebid.server.util.HttpUtil;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class ApplogyBidder implements Bidder<BidRequest> {
+
+    private final String endpointUrl;
+
+    private static final String DEFAULT_BID_CURRENCY = "USD";
+
+    public ApplogyBidder(String endpointUrl) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        final List<Imp> validImps = new ArrayList<>();
+        for (Imp imp : request.getImp()) {
+            if (imp.getBanner() == null && imp.getVideo() == null && imp.getXNative() == null) {
+                errors.add(BidderError.badInput(
+                        String.format("Applogy only supports banner, video and native media types. Ignoring imp id=%s",
+                                imp.getId())));
+                continue;
+            }
+            validImps.add(processImp(imp));
+        }
+
+        if (validImps.isEmpty()) {
+            errors.add(BidderError.badInput("No valid impressions in the bid request"));
+            return Result.of(Collections.emptyList(), errors);
+        }
+
+        final ExtImpApplogy firstImpExt;
+        try {
+            firstImpExt = parseAndValidateImpExt(validImps.get(0));
+        } catch (PreBidException e) {
+            return Result.emptyWithError(BidderError.badInput(e.getMessage()));
+        }
+
+        final BidRequest outgoingRequest = request.toBuilder().imp(validImps).build();
+        final String body = Json.encode(outgoingRequest);
+        final String requestUrl = endpointUrl + "/" + firstImpExt.getToken();
+        final MultiMap headers = resolveHeaders();
+
+        return Result.of(Collections.singletonList(
+                HttpRequest.<BidRequest>builder()
+                        .method(HttpMethod.POST)
+                        .uri(requestUrl)
+                        .headers(headers)
+                        .payload(outgoingRequest)
+                        .body(body)
+                        .build()),
+                errors);
+    }
+
+    private MultiMap resolveHeaders() {
+        return HttpUtil.headers();
+    }
+
+    private ExtImpApplogy parseAndValidateImpExt(Imp imp) {
```

Comment:
As I've previously noted - pass only `ObjectNode ext` from `imp.getExt()`. You don't need other `imp` fields here.